File: content/projectilewars/panorama/scripts/custom_game/vectors.ts
class Vector {
  x: number;
  y: number;
  z: number;

  constructor(x:number,y:number,z:number) {
    this.x = x || 0
    this.y = y || 0
    this.z = z || 0
  }

  toString(): string {
    return "Vector(" + this.x + ", " + this.y + ", " + this.z + ")";
  }
  Length(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
  }
  LengthSquared(): number {
    return this.x * this.x + this.y * this.y + this.z * this.z;
  }
  DistanceTo(v2:Vector): number {
    return Math.sqrt( (v2.x-this.x)*(v2.x-this.x) + (v2.y-this.y)*(v2.y-this.y) + (v2.z-this.z)*(v2.z-this.z) );
  }
  Substract(v2:Vector) {
    return new Vector( this.x - v2.x, this.y - v2.y, this.z - v2.z );
  }
  Add(v2:Vector) {
    return new Vector( this.x + v2.x, this.y + v2.y, this.z + v2.z );
  }
  Scale(s:number) {
    return new Vector( this.x * s, this.y * s, this.z * s );
  }
  ScaleTo(s:number) {
    let length = this.Length();
    if (length == 0){
      return new Vector( 0, 0, 0 ); 
    }
    else {
      return this.Scale( s / length );  
    }
  }
  Normalize() {
    let length = this.Length();
    return new Vector( this.x / length, this.y / length, this.z / length );
  }
  Dot(v2:Vector) {
    return this.x * v2.x + this.y * v2.y + this.z * v2.z;
  }
  Cross(v2:Vector) {
    return new Vector(this.y * v2.z - this.z * v2.y, this.z * v2.x - this.x * v2.z, this.x * v2.y - this.y * v2.x);
  }
}

function ArrayToVector(array:Array<number>):Vector {
  //if (!array) { return new Vector(0, 0, 0);     
  if (array.length == 2) {
    return new Vector( array[0], array[1], 0);
  } else if (array.length == 3) {
    return new Vector( array[0], array[1], array[2]);
  } else {
    return new Vector(0,0,0);
  }
}

function VectorToArray(v:Vector):[number,number,number] {
  //if (!v == undefined || !v.x) {return [0,0,0];}
  let a = v.x || 0
  let b = v.y || 0
  let c = v.z || 0
  return [a,b,c]
}

File: content/projectilewars/panorama/scripts/custom_game/vectors.lua
require("typescript_lualib")
Vector = Vector or {}
Vector.__index = Vector
function Vector.new(construct, ...)
    local instance = setmetatable({}, Vector)
    if construct and Vector.constructor then Vector.constructor(instance, ...) end
    return instance
end
function Vector.constructor(self,x,y,z)
    self.x=(x or 0)
    self.y=(y or 0)
    self.z=(z or 0)
end
function Vector.toString(self)
    return "Vector("..self.x..", "..self.y..", "..self.z..")"
end
function Vector.Length(self)
    return math.sqrt(((self.x*self.x)+(self.y*self.y))+(self.z*self.z))
end
function Vector.LengthSquared(self)
    return ((self.x*self.x)+(self.y*self.y))+(self.z*self.z)
end
function Vector.DistanceTo(self,v2)
    return math.sqrt((((v2.x-self.x)*(v2.x-self.x))+((v2.y-self.y)*(v2.y-self.y)))+((v2.z-self.z)*(v2.z-self.z)))
end
function Vector.Substract(self,v2)
    return Vector.new(true,self.x-v2.x,self.y-v2.y,self.z-v2.z)
end
function Vector.Add(self,v2)
    return Vector.new(true,self.x+v2.x,self.y+v2.y,self.z+v2.z)
end
function Vector.Scale(self,s)
    return Vector.new(true,self.x*s,self.y*s,self.z*s)
end
function Vector.ScaleTo(self,s)
    local length = Vector.Length(self)

    if length==0 then
        return Vector.new(true,0,0,0)
    else
        return Vector.Scale(self,s/length)
    end
end
function Vector.Normalize(self)
    local length = Vector.Length(self)

    return Vector.new(true,self.x/length,self.y/length,self.z/length)
end
function Vector.Dot(self,v2)
    return ((self.x*v2.x)+(self.y*v2.y))+(self.z*v2.z)
end
function Vector.Cross(self,v2)
    return Vector.new(true,(self.y*v2.z)-(self.z*v2.y),(self.z*v2.x)-(self.x*v2.z),(self.x*v2.y)-(self.y*v2.x))
end
function ArrayToVector(array)
    if #array==2 then
        return Vector.new(true,array[0+1],array[1+1],0)
    else
        if #array==3 then
            return Vector.new(true,array[0+1],array[1+1],array[2+1])
        else
            return Vector.new(true,0,0,0)
        end
    end
end
function VectorToArray(v)
    local a = v.x or 0

    local b = v.y or 0

    local c = v.z or 0

    return {a,b,c}
end
